### PR TITLE
Terminate interrupted streams

### PR DESCRIPTION
**Pull request by Lirianer**
Additions:
- New Algora.Terminate GenServer:
  - Looks for videos with duration 0 and is_live false
  - Calls Algora.Library.terminate_stream with the video id
  - Runs every 10 seconds

- New Algora.Library.terminate_stream/1
  - Gets video manifest
  - Appends "#EXT-X-ENDLIST\n" if necessary
  - Updates video duration
  
 Note:
 With this changes Algora.Terminate should not get the same videos again as the duration is updated.
 Although if the duration is less than a second this could be a problem.
 
 Im learning elixir so any suggestions/changes are appreciated.
 
 /claim #69

### CODE REVIEW

File: lib/algora/terminate.ex [hunk 1]
```diff
@@ -0,0 +1,34 @@
+defmodule Algora.Terminate do
+  use GenServer
+
+  import Ecto.Query, warn: false
+
+  @terminate_interval :timer.se(10)
```
Comment (zcesur) [inline lib/algora/terminate.ex:6]: Typo here, let's also increase the interval to 1 hour

File: lib/algora/terminate.ex [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_info(:terminate, state) do
+    from(v in Algora.Library.Video,
+          where: v.duration == 0 and v.is_live == false,
+          select: v.id
+        )
+        |> Algora.Repo.Local.all
```
Comment (zcesur) [inline lib/algora/terminate.ex:24]: `Local` is not needed

File: lib/algora/terminate.ex [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  @impl true
+  def init(state) do
+    schedule_terminate()
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_info(:terminate, state) do
+    from(v in Algora.Library.Video,
+          where: v.duration == 0 and v.is_live == false,
```
Comment (zcesur) [inline lib/algora/terminate.ex:21]: Let's also add a `and v.format == :hls` filter here

File: lib/algora/terminate.ex [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+          select: v.id
+        )
+        |> Algora.Repo.Local.all
+        |> Enum.each(&Algora.Library.terminate_stream/1)
```
Comment (zcesur) [inline lib/algora/terminate.ex:25]: Let's move this chunk into a function like `Library.terminate_interrupted_streams()`

File: lib/algora/library.ex [hunk 5]
```diff
@@ -226,6 +226,29 @@ defmodule Algora.Library do
     |> Repo.one()
   end
 
+  def terminate_stream(video_id) do
+    video = from(v in Video,
+      where: v.id == ^video_id
+    )
+    |> Repo.one()
+
+    {:ok, response} = Finch.build(:get, "#{video.url_root}/g3cFdmlkZW8.m3u8")
```
Comment (zcesur) [inline lib/algora/library.ex:235]: Sometimes this file doesn't exist which causes `get_duration` to crash. Can we add a new field `corrupted` on videos and mark those as such so we can filter them out?

**Review by zcesur — COMMENTED** [review]
Looks good overall, thanks for contributing! Needs a few minor changes before we merge it in

How do you like Elixir so far? :)

File: lib/algora/terminate.ex (hunk 1 above)
Comment (Lirianer) [reply lib/algora/terminate.ex]: Whoops, done! New timer is for 1 hour

**Comment by Lirianer** [comment]
I've done the changes you pointed out!

It's been a blast so far, although I do waste a lot of time thinking on  whether something should be a new function or not and if im missing a better/more idiomatic way of doing things

**Comment by zcesur** [comment]
Awesome, thanks!

I converted `case` into `with` and moved `get_duration` up there so we handle any errors there as well (233a53d)

We can proceed to merge & reward once you sign our CLA

**Comment by zcesur** [comment]
If you're on Twitter I'd love to tweet about you being our first contributor :)

Btw feel free to DM me any time or use the #contributing channel in our [Discord](https://algora.io/discord) for anything Algora/Elixir related

**Comment by Lirianer** [comment]
Sure! I'm [@Lirianer_](https://x.com/Lirianer_) on Twitter/X.

Thanks, just joined the discord!

**Pull request merged**